### ISSUE TITLE
FAILED SystemMessageDeliveryNoPassiveRetryGate

### ISSUE THREAD

**Issue by jrudolph** [akka/akka#21684]
Oct 10, 2016 akka-multi-node-nightly2 https://jenkins.akka.io:8498/job/akka-multi-node-nightly2/506/consoleFull

```
[info] SystemMessageDeliveryNoPassiveRetryGate:
Coroner Thread Count starts at 109 in akka.remote.transport.SystemMessageDeliveryNoPassiveRetryGate
[info] Remoting passive connections off
[WARN] [10/17/2016 08:17:36.959] [pool-42-thread-10-ScalaTest-running-SystemMessageDeliveryNoPassiveRetryGate] [akka.serialization.Serialization(akka://systemB)] Using the default Java serializer for class [scala.concurrent.duration.FiniteDuration] which is not recommended because of performance implications. Use another serializer or disable this warning using the setting 'akka.actor.warn-about-java-serializer-usage'
[WARN] [10/17/2016 08:17:36.959] [pool-42-thread-10-ScalaTest-running-SystemMessageDeliveryNoPassiveRetryGate] [akka.serialization.Serialization(akka://SystemMessageDeliveryStressTest)] Using the default Java serializer for class [scala.concurrent.duration.FiniteDuration] which is not recommended because of performance implications. Use another serializer or disable this warning using the setting 'akka.actor.warn-about-java-serializer-usage'
[WARN] [10/17/2016 08:17:36.967] [systemB-akka.remote.default-remote-dispatcher-8] [akka.gremlin.trttl.tcp://systemB@127.0.1.1:33576/system/endpointManager/reliableEndpointWriter-akka.gremlin.trttl.tcp%3A%2F%2FSystemMessageDeliveryStressTest%40127.0.1.1%3A49900-0] Association with remote system [akka.gremlin.trttl.tcp://SystemMessageDeliveryStressTest@127.0.1.1:49900] has failed, address is now gated for [500] ms. Reason: [Association failed with [akka.gremlin.trttl.tcp://SystemMessageDeliveryStressTest@127.0.1.1:49900]] Caused by: [Simulated failure of association to trttl.tcp://SystemMessageDeliveryStressTest@127.0.1.1:49900]
[WARN] [10/17/2016 08:17:37.488] [SystemMessageDeliveryStressTest-akka.actor.default-dispatcher-3] [akka://SystemMessageDeliveryStressTest/system/endpointManager/reliableEndpointWriter-akka.gremlin.trttl.tcp%3A%2F%2FsystemB%40127.0.1.1%3A33576-0] unhandled message from Actor[akka://SystemMessageDeliveryStressTest/system/endpointManager#585544004]: Ungate
[info] - must guaranteed delivery and message ordering despite packet loss  *** FAILED *** (536 milliseconds)
[info]   java.lang.AssertionError: assertion failed: expected 0, found 100
[info]   at scala.Predef$.assert(Predef.scala:170)
[info]   at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:388)
[info]   at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:373)
[info]   at akka.testkit.TestKit.expectMsg(TestKit.scala:814)
[info]   at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$4.apply$mcVI$sp(SystemMessageDeliveryStressTest.scala:172)
[info]   at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
[info]   at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(SystemMessageDeliveryStressTest.scala:169)
[info]   at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(SystemMessageDeliveryStressTest.scala:145)
[info]   at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(SystemMessageDeliveryStressTest.scala:145)
[info]   at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.WordSpecLike$$anon$1.apply(WordSpecLike.scala:1078)
[info]   at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
[info]   at akka.testkit.AkkaSpec.withFixture(AkkaSpec.scala:57)
[info]   at org.scalatest.WordSpecLike$class.invokeWithFixture$1(WordSpecLike.scala:1075)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTest$1.apply(WordSpecLike.scala:1088)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
[info]   at org.scalatest.WordSpecLike$class.runTest(WordSpecLike.scala:1088)
[info]   at akka.testkit.AkkaSpec.runTest(AkkaSpec.scala:57)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.WordSpecLike$$anonfun$runTests$1.apply(WordSpecLike.scala:1147)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:381)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
[info]   at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
[info]   at scala.collection.immutable.List.foreach(List.scala:381)
[info]   at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
[info]   at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
[info]   at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
[info]   at org.scalatest.WordSpecLike$class.runTests(WordSpecLike.scala:1147)
[info]   at akka.testkit.AkkaSpec.runTests(AkkaSpec.scala:57)
[info]   at org.scalatest.Suite$class.run(Suite.scala:1147)
[info]   at akka.testkit.AkkaSpec.org$scalatest$WordSpecLike$$super$run(AkkaSpec.scala:57)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.WordSpecLike$$anonfun$run$1.apply(WordSpecLike.scala:1192)
[info]   at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
[info]   at org.scalatest.WordSpecLike$class.run(WordSpecLike.scala:1192)
[info]   at akka.testkit.AkkaSpec.org$scalatest$BeforeAndAfterAll$$super$run(AkkaSpec.scala:57)
[info]   at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
[info]   at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
[info]   at akka.testkit.AkkaSpec.run(AkkaSpec.scala:57)
[info]   at org.scalatest.tools.Framework.org$scalatest$tools$Framework$$runSuite(Framework.scala:314)
[info]   at org.scalatest.tools.Framework$ScalaTestTask.execute(Framework.scala:472)
[info]   at sbt.TestRunner.runTest$1(TestFramework.scala:76)
[info]   at sbt.TestRunner.run(TestFramework.scala:85)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[info]   at sbt.TestFunction.apply(TestFramework.scala:207)
[info]   at sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[info]   at sbt.std.Transform$$anon$4.work(System.scala:63)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[info]   at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[info]   at sbt.Execute.work(Execute.scala:237)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[info]   at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
[info]   at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[info]   at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[info]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[info]   at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[info]   at java.lang.Thread.run(Thread.java:745)
```

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/3856/consoleFull

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/3882/consoleFull

**Comment by johanandren** [comment]
Researched that date but cannot find anything that seem relevant getting merged (the tls docs, remote deploy whitelist are the closest it gets).

That it expects 0 but gets 100 means that the entire first burst of messages got lost (it sends 5000 messages in bursts of 100), or possibly that the second burst arrived before the first.

**Comment by 2m** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-2.12/36/consoleFull

**Comment by patriknw** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/3937/consoleFull

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-2.12/47/consoleFull

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-nightly-2.12/70/consoleFull

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-release-2.4/77/console

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/1929/console

**Comment by drewhk** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/4144/consoleFull

**Comment by 2m** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4667/consoleFull

**Comment by 2m** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4679/consoleFull

**Comment by ktoso** [comment]
Again https://jenkins.akka.io:8498/job/akka-nightly/4195/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/4202/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4706/

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4725/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by bantonsson** [comment]
This seems to be a fairly popular failure. I'll give it a shot.

**Comment by bantonsson** [comment]
After excavating the code, my only idea so far is that there is a race while establishing connections and gating, resulting in some kind of uid skew that results in a loss of data in the saved resend state or a reset in the outgoing acked buffer. I increased the likelihood of dropping the connection and gating and have been running this on repeat over the weekend. Now after 700+ runs, there hasn't been a single failure 😞 I'll put this one back and pick up another one.

**Comment by jrudolph** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4824/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by ktoso** [comment]
https://jenkins.akka.io:8498/job/akka-nightly/4391/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

```
Error Message

assertion failed: expected 0, found 100
Stacktrace

      java.lang.AssertionError: assertion failed: expected 0, found 100
      at scala.Predef$.assert(Predef.scala:170)
      at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:387)
      at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:372)
      at akka.testkit.TestKit.expectMsg(TestKit.scala:828)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$4.apply$mcVI$sp(SystemMessageDeliveryStressTest.scala:172)
      at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(SystemMessageDeliveryStressTest.scala:170)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(SystemMessageDeliveryStressTest.scala:146)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$ano
```

**Comment by raboof** [comment]
https://jenkins.akka.io:8498/job/akka-release-2.4/212/testReport/junit/akka.remote.transport/SystemMessageDeliveryRetryGate/Remoting_passive_connections_on_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/
```
java.lang.AssertionError: assertion failed: expected 0, found 100
      at scala.Predef$.assert(Predef.scala:170)
      at akka.testkit.TestKitBase$class.expectMsg_internal(TestKit.scala:388)
      at akka.testkit.TestKitBase$class.expectMsg(TestKit.scala:373)
      at akka.testkit.TestKit.expectMsg(TestKit.scala:814)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1$$anonfun$apply$mcV$sp$4.apply$mcVI$sp(SystemMessageDeliveryStressTest.scala:172)
      at scala.collection.immutable.Range.foreach$mVc$sp(Range.scala:160)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply$mcV$sp(SystemMessageDeliveryStressTest.scala:169)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(SystemMessageDeliveryStressTest.scala:145)
      at akka.remote.transport.SystemMessageDeliveryStressTest$$anonfun$1$$anonfun$apply$mcV$sp$1.apply(SystemMessageDeliveryStressTest.scala:145)
```

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node-nightly/4965/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by johanandren** [comment]
https://jenkins.akka.io:8498/job/akka-multi-node/2133/testReport/junit/akka.remote.transport/SystemMessageDeliveryNoPassiveRetryGate/Remoting_passive_connections_off_must_guaranteed_delivery_and_message_ordering_despite_packet_loss/

**Comment by johanandren** [comment]
Hmm, this hasn't happened for half a year, I wonder if we fixed something. Possibly in the fix for #22156

**Comment by johanandren** [comment]
Closing since it hasn't happened in more than a year.
